### ISSUE TITLE
Incorrect mIoU calculation?

### ISSUE THREAD

**Issue by vikolss** [yassouali/pytorch_segmentation#58]
Hi, thank you for the very well written code! 
I have found an issue however, or actually two issues.
Firstly, when calculating mIoU, you add the sizes of intersections from all samples, and add the unions from all samples, and then divide the two. Wouldn't the correct way be to calculate IoU for each sample and then add these? These are not equal as the mean of quotients is not equal to the quotient of sums.
Secondly, when calculating the mean IoU over classes you include the classes that did not have any instances in either the prediction or the ground truth, i.e. where both the intersection and the union is zero. These will significantly lower the mIoU in a way it should not, right?

The first issue somewhat neutralises the second, as most (or all) classes will be represented when adding a lot of samples together. The result should still be erroneous though, which could be part of the reason for people not being able to recreate results from papers?

Please let me know if you agree with these issues, or if I have misunderstood something.

**Comment by yassouali** [comment]
Hi, thanks for the question.

For the first issue, computing the mIoU over all examples and each one separately does give the same results, maybe I didn't understand your issue quite right.

For the second issue, you're right, at the beginning of training (or eval) the mIoU will not be correct, however, the final value (or after some seeing batches) will be since we'll have all of the classes.

And i I remember correctly, I did compare the mIoU results with the ones computed using chainer's `eval_semantic_segmentation` and I got the same results.

**Comment by vikolss** [comment]
Thanks for the quick reply.

The included image illustrates what I mean with the quotients not being equal. After some investigation I have found that depending on the sizes of the individual IoU scores (i.e. the performance of the network) the scores differ in different ways. When individual IoUs on average are larger than 0.5, the right hand side in the inequality (your method) is larger, while when IoUs on average are smaller than 0.5, the left hand side is instead larger. (I tried this for many samples, not just two as in the inequality). 

Which method is correct for calculating mIoU in semantic segmentation I am not sure of however. I think that averaging over IoUs feels more intuitive, but I have seen the method you use in several other implementations, so that may very well be the correct way.

For the second issue I agree that it is not important when calculating mIoU the way you do it.

![latex](https://user-images.githubusercontent.com/35954407/76395721-3edb3480-6378-11ea-8dcf-c24bab489cf6.jpg)

**Comment by yassouali** [comment]
Oh ok, I see what you mean, but I do think the correct way to do it is over the whole images, this way the results are more representative of the real performance of the model, where, if we summed the mIoUs of each image, some might be quite trivial (eg, majority of image is background or one class) and might slightly skew the results (they will be quite similar tho).

**Comment by yassouali** [comment]
If you still have some doubts we can repoen the issue

**Comment by sanje2v** [comment]
If we separate (a+b)/(c+d) into (a/(c+d)) + (b/(c+d)), we can see that 'a' is divided also by 'd' to which is has no relation to (as it would come from a different image) and similarly for 'b' with 'c'. Also, as the metric is called 'mean' IoU, so the '0.5 * (a/c + b/d)' seems to the correct method.

Did you find an answer to this @vikolss ?

**Comment by vikolss** [comment]
In my opinion '0.5 * (a/c + b/d)' is a more intuitive way of calculating a mean intersection over union. But the way that it is done in this repository, (a+b)/(c+d), is the same way it is done in several other repositories that I have looked at. So I believe that that is the correct way to do it.
